ygot/struct_validation_map.go

```diff
@@ -666,23 +666,53 @@ func copyInterfaceField(dstField, srcField reflect.Value, opts ...MergeOpt) erro
 		return nil
 	}
 
-	if !util.IsValueInterface(srcField) || !util.IsValueStructPtr(srcField.Elem()) {
-		return fmt.Errorf("invalid interface type received: %T", srcField.Interface())
-	}
-	s := srcField.Elem().Elem() // Dereference src to a struct.
-	if !util.IsNilOrInvalidValue(dstField) {
-		dV := dstField.Elem().Elem() // Dereference dst to a struct.
-		if !fieldOverwriteEnabled(opts) && !reflect.DeepEqual(s.Interface(), dV.Interface()) {
-			return fmt.Errorf("interface field was set in both src and dst and was not equal, src: %v, dst: %v", s.Interface(), dV.Interface())
+	if !util.IsValueInterface(srcField) {
+		return fmt.Errorf("non-interface type received: %T", srcField.Interface())
+	}
+
+	_, isGoEnum := srcField.Elem().Interface().(GoEnum)
+	switch {
+	case util.IsValueStructPtr(srcField.Elem()):
+		s := srcField.Elem().Elem() // Dereference src to a struct.
+		if !util.IsNilOrInvalidValue(dstField) {
+			dV := dstField.Elem().Elem() // Dereference dst to a struct.
+			if diff := cmp.Diff(s.Interface(), dV.Interface()); !fieldOverwriteEnabled(opts) && diff != "" {
+				return fmt.Errorf("interface field was set in both src and dst and was not equal, (-src, +dst):\n%s", diff)
+			}
 		}
-	}
 
-	d := reflect.New(s.Type())
-	if err := copyStruct(d.Elem(), s, opts...); err != nil {
-		return err
+		d := reflect.New(s.Type())
+		if err := copyStruct(d.Elem(), s, opts...); err != nil {
+			return err
+		}
+		dstField.Set(d)
+		return nil
+	case srcField.Elem().Kind() == reflect.Slice && srcField.Elem().Type().Name() == BinaryTypeName:
+		if !util.IsNilOrInvalidValue(dstField) {
+			s, d := srcField.Interface(), dstField.Interface()
+			if diff := cmp.Diff(s, d); !fieldOverwriteEnabled(opts) && diff != "" {
+				return fmt.Errorf("interface field was set in both src and dst and was not equal, (-src, +dst):\n%s", diff)
+			}
+		}
+
+		srcVal := srcField.Elem()
+		ns := reflect.Zero(srcVal.Type())
+		for i := 0; i < srcVal.Len(); i++ {
+			ns = reflect.Append(ns, srcVal.Index(i))
+		}
+		dstField.Set(ns)
+		return nil
+	case util.IsValueScalar(srcField.Elem()) && (isGoEnum || unionTypedefUnderlyingTypes[srcField.Elem().Type().Name()] != nil):
```

Comment:
Should this be declared in `ygen` since it's really ygen that defines what can be output here, rather than ygot?

Reply:
That's not possible because `ygen` depends on `ygot`. There is necessarily some type name information that `ygot` needs to know, e.g. `BinaryTypeName`, so I think it's reasonable to leave it in `ygot.